src/sort.rs

```diff
@@ -302,127 +237,73 @@ where
-                    end_l = end_l.offset(!is_less(&*elem, pivot) as isize);
-                    elem = elem.offset(1);
-                }
+                offsets_l[end_l] = i as u8;
+                end_l += !is_less(&v[elem], pivot) as usize;
+                elem += 1;
             }
         }
 
         if start_r == end_r {
             // Trace `block_r` elements from the right side.
-            start_r = offsets_r.as_mut_ptr();
-            end_r = offsets_r.as_mut_ptr();
+            start_r = 0;
+            end_r = 0;
             let mut elem = r;
 
             for i in 0..block_r {
-                // SAFETY: The unsafety operations below involve the usage of the `offset`.
-                //         According to the conditions required by the function, we satisfy them because:
-                //         1. `offsets_r` is stack-allocated, and thus considered separate allocated object.
-                //         2. The function `is_less` returns a `bool`.
-                //            Casting a `bool` will never overflow `isize`.
-                //         3. We have guaranteed that `block_r` will be `<= BLOCK`.
-                //            Plus, `end_r` was initially set to the begin pointer of `offsets_` which was declared on the stack.
-                //            Thus, we know that even in the worst case (all invocations of `is_less` returns true) we will only be at most 1 byte pass the end.
-                //        Another unsafety operation here is dereferencing `elem`.
-                //        However, `elem` was initially `1 * sizeof(T)` past the end and we decrement it by `1 * sizeof(T)` before accessing it.
-                //        Plus, `block_r` was asserted to be less than `BLOCK` and `elem` will therefore at most be pointing to the beginning of the slice.
-                unsafe {
-                    // Branchless comparison.
-                    elem = elem.offset(-1);
-                    *end_r = i as u8;
-                    end_r = end_r.offset(is_less(&*elem, pivot) as isize);
-                }
+                elem -= 1;
+                offsets_r[end_r] = i as u8;
+                end_r += is_less(&v[elem], pivot) as usize;
             }
         }
 
         // Number of out-of-order elements to swap between the left and right side.
         let count = cmp::min(width(start_l, end_l), width(start_r, end_r));
 
         if count > 0 {
-            macro_rules! left {
-                () => {
-                    l.offset(*start_l as isize)
-                };
-            }
-            macro_rules! right {
-                () => {
-                    r.offset(-(*start_r as isize) - 1)
-                };
-            }
-
-            // Instead of swapping one pair at the time, it is more efficient to perform a cyclic
-            // permutation. This is not strictly equivalent to swapping, but produces a similar
-            // result using fewer memory operations.
```

Comment:
This comment should be retained.

Reply:
It isn't actually true for the replacement code (which performs the same number of memory operations as naive swapping would).